### PR TITLE
Enable loading snapshots with sound null safety enabled.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Snapshots compiled with sound null-safety enabled require changes to the way in
which isolates are launched. Specifically, the `Dart_IsolateFlags::null_safety`
field needs to be known upfront. The value of this field can only be determined
once the kernel snapshot is available. This poses a problem in the engine
because the engine used to launch the isolate at shell initialization and only
need the kernel mappings later at isolate launch (when transitioning the root
isolate to the `DartIsolate::Phase::Running` phase). This patch delays launch of
the isolate on the UI task runner till a kernel mapping is available. The side
effects of this delay (callers no longer having access to the non-running
isolate handle) have been addressed in this patch. The DartIsolate API has also
been amended to hide the method that could return a non-running isolate to the
caller.  Instead, it has been replaced with a method that requires a valid
isolate configuration that returns a running root isolate. The isolate will be
launched by asking the isolate configuration for its null-safety
characteristics.

A side effect of enabling null-safety is that Dart APIs that work with legacy
types will now terminate the process if used with an isolate that has sound
null-safety enabled. These APIs may no longer be used in the engine. This
primarily affects the Dart Convertors in Tonic that convert certain C++ objects
into the Dart counterparts. All known Dart Converters have been updated to
convert C++ objects to non-nullable Dart types inferred using type traits of the
corresponding C++ object. The few spots in the engine that used the old Dart
APIs directly have been manually updated. To ensure that no usage of the legacy
APIs remain in the engine (as these would cause runtime process terminations),
the legacy APIs were prefixed with the `DART_LEGACY_API` macro and the macro
defined to `[[deprecated]]` in all engine translation units. While the engine
now primarily works with non-nullable Dart types, callers can still use
`Dart_TypeToNonNullableType` to acquire nullable types for use directly or with
Tonic. One use-case that is not addressed with the Tonic Dart Convertors is the
creation of non-nullable lists of nullable types. This hasn’t come up so far in
the engine.

A minor related change is reworking tonic to define a single library target.
This allows the various tonic subsystems to depend on one another. Primarily,
this is used to make the Dart convertors use the logging utilities. This now
allows errors to be more descriptive as the presence of error handles is caught
(and logged) earlier.

Fixes https://github.com/flutter/flutter/issues/59879

### CODE REVIEW

**Comment by chinmaygarde** [comment]
The failing test in the presubmits [was added 4 days ago](https://github.com/flutter/engine/pull/21737) and deals with with the nullability requirement for the main dart entrypoints arguments. @a-siva is clarifying the spec.

**Comment by chinmaygarde** [comment]
Ok, fixed the [behavior to match the clarification about `main` for null-safety](https://github.com/dart-lang/language/pull/1146). Existing test should match spec.

File: runtime/dart_lifecycle_unittests.cc
Comment (zanderso) [inline runtime/dart_lifecycle_unittests.cc:4]: We're supposed to be trying to get rid of these =)

File: runtime/dart_snapshot.cc
Comment (zanderso) [inline runtime/dart_snapshot.cc:207]: Earlier you used `std::optional`. How are you choosing between that and this?

File: runtime/type_conversions_unittests.cc
Comment (zanderso) [inline runtime/type_conversions_unittests.cc:52]: Should that go in this PR, or should we file an issue for it?

File: runtime/dart_lifecycle_unittests.cc
Comment (zanderso) [reply runtime/dart_lifecycle_unittests.cc:4]: Oh, or maybe that's why you hit the linter bugs...

Comment (chinmaygarde) [reply runtime/dart_lifecycle_unittests.cc:4]: Yeah, since I touched the file, the linter linted the whole file. Didn't think reworking the APIs to make the linter happy was in scope for this patch.

File: runtime/dart_snapshot.cc
Comment (chinmaygarde) [reply runtime/dart_snapshot.cc:207]: IIRC, earlier it was an optional vector, now its an pointer that may be null. I could make it an optional pointer or an optional reference. A pointer just seemed cleaner here.

File: runtime/type_conversions_unittests.cc
Comment (chinmaygarde) [reply runtime/type_conversions_unittests.cc:52]: Oops. I thought I cleaned up all `TODO(WIP)`s. No, I'll add the tests as there are no tests in Tonic for the convertors I touched.

Comment (chinmaygarde) [reply runtime/type_conversions_unittests.cc]: Done. [Added the tests](https://github.com/flutter/engine/pull/21820/commits/eb4f87627f51959ce5491d2cb1db5f3276e33ca7) meant for this harness. Also checked again to make sure I didn't miss any WIP items.

File: testing/testing.gni
Comment (zanderso) [inline testing/testing.gni:83]: Does this mean that all of the Dart language tests will run in nnbd mode?

**Review by zanderso — APPROVED** [review]
lgtm w/ nits and questions. @jason-simmons should also review.

File: common/config.gni
Comment (chinmaygarde) [reply common/config.gni:31]: As I said in the commit message, the way we ensure that no APIs that work with legacy are used in our codebase is to decorate these APIs with the `DART_LEGACY_API` macro and then replace that with `[[deprecated]]` macro at runtime. [This patch in the Dart VM](https://dart-review.googlesource.com/c/sdk/+/167542) adds these decorations. I am hoping to land this in the VM as a safeguard against unsafe uses of the legacy API creep into our codebase. This is dangerous and these APIs will cause process terminations at runtime (and only when sound null-safety is enabled). If I cannot land that Dart VM patch or we decide these safeguards are not necessary, I can back this out.

File: runtime/dart_isolate.cc
Comment (chinmaygarde) [reply runtime/dart_isolate.cc]: Yes, I missed "shutdown". Added.

File: runtime/dart_isolate.h
Comment (chinmaygarde) [reply runtime/dart_isolate.h]: Using `explicit` for single argument constructors is an explicit recommendation (https://google.github.io/styleguide/cppguide.html#Implicit_Conversions) but I didn't think there was anything against `explicit` for other constructors. I don't this this is necessary here. I'll get rid of it.

File: testing/dart_isolate_runner.cc
Comment (chinmaygarde) [reply testing/dart_isolate_runner.cc:132]: You mean just return `std::shared_ptr`. I don't see why not. I like `weak_ptr` because the docs explicitly mention not to hold onto strong references to the isolate as that can interfere with the object from being collected if the isolate decides to kill itself. But if you have a strong opinion about this, I can update all callsites.

File: testing/testing.gni
Comment (chinmaygarde) [reply testing/testing.gni:83]: This is used by the engine unit-tests that have Dart fixtures. The Dart unit-tests still run with the default snapshot generation steps (which don't use strong null-safety). I don't think the snaphotting for the scenarios app uses strong null-safety either.

File: common/config.gni
Comment (zanderso) [reply common/config.gni:31]: Got it. Commented on that patch.

File: testing/dart_isolate_runner.cc
Comment (zanderso) [reply testing/dart_isolate_runner.cc:132]: If forgetting the `lock()` call could cause a bug, then I'd change it. If that is unlikely, then I don't have a strong opinion.

Comment (chinmaygarde) [reply testing/dart_isolate_runner.cc:132]: Forgetting a lock would make the variable a weak pointer and that doesn't have a dereference operator. Calling any methods on the isolate would be impossible.

**Comment by chinmaygarde** [comment]
The Fuchsia presubmit [is this flake](https://github.com/flutter/flutter/issues/67942). The release mode test failure seems real, investigating.

**Review by a-siva — APPROVED** [review]
The command line VM also has a command line flag --sound-null-safety which is used to override the null safety mode, I presume the engine would also want something like that to override whatever is specified in the snapshot or kernel file?
Not sure if users would want it but something to consider.

File: third_party/tonic/converter/dart_converter.h
Comment (chinmaygarde) [reply third_party/tonic/converter/dart_converter.h:329]: > It doesn't look like CreateZeroInitializedDartObject is ever passed a null object as the type.

It is used above in [`//fml/dart/dart_converter.h`](https://github.com/flutter/engine/pull/21820/files#diff-30826356f1934632f024ffea52352e2d42840066513868f17651fccfc1c7a4cfR28).

`type_handle_or_null` is only `null` when the type can be matched by one of the type traits. In reality, it can always be `null` because we can use `ToDartTypeHandle` here to get the handle. But I left this in as an optimization because callers always have a handle to the type (like in the case of lists). And, in that case, I didn't want to resolve the library and lookup the type in it again. I can get rid of the argument if it makes things easier to read.

File: third_party/tonic/converter/dart_converter.h
Comment (a-siva) [inline third_party/tonic/converter/dart_converter.h:318]: Not sure how frequently ToDartTypeHandle() is called for these basic types (string, int, double), depending on that it might make sense to cache these types and avoid the need to lookup the library and get the non nullable type everytime.

File: third_party/tonic/converter/dart_converter.h
Comment (chinmaygarde) [reply third_party/tonic/converter/dart_converter.h:357]: Yes, this code is unreachable. The engine doesn't have a macro for unreachable code. [I've filed an issue for it and referenced this comment](https://github.com/flutter/flutter/issues/68164). I'll come patch this site with the new macro when it is in place.

**Review by chinmaygarde — COMMENTED** [review]
The issue in the release mode test is a race that I am still tracking down (doesn't seem to be a data race).

File: third_party/tonic/converter/dart_converter.h
Comment (a-siva) [reply third_party/tonic/converter/dart_converter.h:318]: Maybe not for this PR but something to consider for future.

File: third_party/tonic/converter/dart_converter.h
Comment (a-siva) [reply third_party/tonic/converter/dart_converter.h:329]: Ok, I understand it will be null for string, is_integral and is_floating_point.
Can we add an assert in the final else that type_handle_or_null is not ::Dart_Null()

**Comment by chinmaygarde** [comment]
@a-siva I tracked down the cause of the failing unit-test and it makes no sense to me. Pushing the following transform on the Dart side gives me the incorrect `Float64List` on the native side. This happens only in release mode. That is why the test assertions are failing. I didn't touch any of the converters for typed list. Not sure what is going with this happening only in release mode. But it is 100% reproducible. Can we pair on figuring this out. I am a bit lost on this one.

In Dart:
```
builder.pushTransform(Float64List.fromList([
      2.123, 0.0, 0.0, 0.0,
      0.0, 2.123, 0.0, 0.0,
      0.0, 0.0, 1.0, 0.0,
      0.0, 0.0, 0.0, 1.0
    ]));
```

In Native (notice the incorrect entry at index 6):
```
ERROR:flutter/lib/ui/compositing/scene_builder.cc(94)] ~~~~~~~~~~~~~~~~~~~~~
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(95)] Pushing transform.
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[0]=2.123
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[1]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[2]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[3]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[4]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[5]=1
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[6]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[7]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[8]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[9]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[10]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[11]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[12]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[13]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[14]=0
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(97)] matrix4[15]=1
[ERROR:flutter/lib/ui/compositing/scene_builder.cc(99)] ~~~~~~~~~~~~~~~~~~~~~
```

**Comment by chinmaygarde** [comment]
Update: Siva and I managed to isolate the issue down to `Float64List.fromList` constructing an incorrect list in AOT mode with sound null-safety.

```
  var list = Float64List.fromList([
      1, 2, 3, 4,
      5, 6, 7, 8,
      9, 10, 11, 12,
      13, 14, 15, 16
    ]);
    print(list);
```

Prints:
```
[1.0, 3.0, 5.0, 7.0, 9.0, 11.0, 13.0, 15.0, 0.0, 10.0, 0.0, 12.0, 0.0, 14.0, 0.0, 16.0]
```

**Comment by chinmaygarde** [comment]
Blocked by https://github.com/dart-lang/sdk/issues/43810.

**Review by jason-simmons — APPROVED** [review]

**Comment by chinmaygarde** [comment]
The [fix in the Dart VM has been identified](https://github.com/dart-lang/sdk/issues/43810#issuecomment-709640639) but [the patch](https://dart-review.googlesource.com/c/sdk/+/167865) not yet reviewed. Testing locally and will rebase once the fix has landed and been rolled into the engine.

**Comment by vsmenon** [comment]
@chinmaygarde - did you manage to test locally?  Do we believe the pending fix is the last item to land this?

**Comment by chinmaygarde** [comment]
The patch has landed and is [waiting to be rolled](https://github.com/flutter/engine/pull/21919) into the engine here. The roll presubmits are ongoing but the tree is red because of [this infra flake (P0)](https://github.com/flutter/flutter/issues/68328).

> did you manage to test locally? Do we believe the pending fix is the last item to land this?
@vsmenon: Yes, patching this in locally resolved the failures. I don't know of any other blockers to landing this patch once the roll is done.

**Comment by chinmaygarde** [comment]
Note: The latest update I pushed only rebases on ToT resolves conflicts. The roll has still not landed. Presubmits will still fail because of the earlier reason.

**Comment by chinmaygarde** [comment]
Rebased after the roll and ran a few smoke tests in the release product mode locally. Seems fine.

**Comment by blasten** [comment]
I was running the hello world app with this engine change, and got:

```
An Observatory debugger and profiler on Android SDK built for x86 64 is available at: http://127.0.0.1:53216/61LLdqm7fvM=/
I/flutter (11150): ══╡ EXCEPTION CAUGHT BY SCHEDULER LIBRARY ╞═════════════════════════════════════════════════════════
I/flutter (11150): The following _TypeError was thrown during a scheduler callback:
I/flutter (11150): type 'Null' is not a subtype of type 'EngineLayer' of 'function result'
I/flutter (11150):
I/flutter (11150): When the exception was thrown, this was the stack:
I/flutter (11150): #0      SceneBuilder._pushPhysicalShape (dart:ui/compositing.dart)
I/flutter (11150): #1      SceneBuilder.pushPhysicalShape (dart:ui/compositing.dart:577:5)
I/flutter (11150): #2      PhysicalModelLayer.addToScene (package:flutter/src/rendering/layer.dart:1988:29)
I/flutter (11150): #3      Layer._addToSceneWithRetainedRendering (package:flutter/src/rendering/layer.dart:440:5)
I/flutter (11150): #4      ContainerLayer.addChildrenToScene (package:flutter/src/rendering/layer.dart:1027:15)
I/flutter (11150): #5      OffsetLayer.addToScene (package:flutter/src/rendering/layer.dart:1164:5)
I/flutter (11150): #6      Layer._addToSceneWithRetainedRendering (package:flutter/src/rendering/layer.dart:440:5)
I/flutter (11150): #7      ContainerLayer.addChildrenToScene (package:flutter/src/rendering/layer.dart:1027:15)
I/flutter (11150): #8      OffsetLayer.addToScene (package:flutter/src/rendering/layer.dart:1164:5)
I/flutter (11150): #9      Layer._addToSceneWithRetainedRendering (package:flutter/src/rendering/layer.dart:440:5)
I/flutter (11150): #10     ContainerLayer.addChildrenToScene (package:flutter/src/rendering/layer.dart:1027:15)
I/flutter (11150): #11     TransformLayer.addToScene (package:flutter/src/rendering/layer.dart:1609:5)
I/flutter (11150): #12     ContainerLayer.buildScene (package:flutter/src/rendering/layer.dart:765:5)
I/flutter (11150): #13     RenderView.compositeFrame (package:flutter/src/rendering/view.dart:231:37)
I/flutter (11150): #14     RendererBinding.drawFrame (package:flutter/src/rendering/binding.dart:447:18)
I/flutter (11150): #15     WidgetsBinding.drawFrame (package:flutter/src/widgets/binding.dart:907:13)
I/flutter (11150): #16     RendererBinding._handlePersistentFrameCallback (package:flutter/src/rendering/binding.dart:309:5)
I/flutter (11150): #17     SchedulerBinding._invokeFrameCallback (package:flutter/src/scheduler/binding.dart:1117:15)
I/flutter (11150): #18     SchedulerBinding.handleDrawFrame (package:flutter/src/scheduler/binding.dart:1055:9)
I/flutter (11150): #19     SchedulerBinding.scheduleWarmUpFrame.<anonymous closure> (package:flutter/src/scheduler/binding.dart:864:7)
I/flutter (11150): (elided 11 frames from class _RawReceivePortImpl, class _Timer, dart:async, and dart:async-patch)
```

App:
```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Welcome to Flutter',
      home: Scaffold(
        appBar: AppBar(
          title: Text('Welcome to Flutter'),
        ),
        body: Center(
          child: Text('Hello World'),
        ),
      ),
    );
  }
}
```

@chinmaygarde  Are you aware of this issue?

**Comment by chinmaygarde** [comment]
This should be fixed [by this patch](https://github.com/flutter/engine/commit/895940081c054f6ce8d5673cb5abf5f2fd83d851).

**Pull request merged**
